python_modules/dagster/dagster/_core/definitions/multi_dimensional_partitions.py

```diff
@@ -246,43 +252,94 @@ def get_multi_partition_key_from_str(self, partition_key_str: str) -> MultiParti
+            for dim in self.partitions_defs
+            if isinstance(dim.partitions_def, TimeWindowPartitionsDefinition)
+        ]
+        if len(time_dimensions) == 1:
+            primary_dimension, secondary_dimension = time_dimensions[0], next(
+                iter([dim for dim in self.partitions_defs if dim != time_dimensions[0]])
+            )
+        else:
+            primary_dimension, secondary_dimension = (
+                self.partitions_defs[0],
+                self.partitions_defs[1],
+            )
+
+        return primary_dimension, secondary_dimension
+
+    @property
+    def primary_dimension(self) -> PartitionDimensionDefinition:
+        return self._get_primary_and_secondary_dimension()[0]
+
+    @property
+    def secondary_dimension(self) -> PartitionDimensionDefinition:
+        return self._get_primary_and_secondary_dimension()[1]
+
 
 class MultiPartitionsSubset(DefaultPartitionsSubset):
-    @staticmethod
-    def from_serialized(
-        partitions_def: PartitionsDefinition, serialized: str
-    ) -> "MultiPartitionsSubset":
-        if not isinstance(partitions_def, MultiPartitionsDefinition):
-            check.failed(
-                "Must pass a MultiPartitionsDefinition object to deserialize MultiPartitionsSubset."
+    def __init__(
+        self,
+        partitions_def: MultiPartitionsDefinition,
+        subset: Optional[Set[str]] = None,
+    ):
+        """
+        Users should not directly call this method. Instead, call partitions_def.empty_subset().with_partition_keys().
+        """
+        check.inst_param(partitions_def, "partitions_def", MultiPartitionsDefinition)
+        subset = (
+            set(partitions_def.get_partition_key_from_str(key) for key in subset)
+            if subset
+            else set()
+        )
+        super(MultiPartitionsSubset, self).__init__(partitions_def, subset)
+
+    def with_partition_keys(self, partition_keys: Iterable[str]) -> "MultiPartitionsSubset":
+        partitions_def = cast(MultiPartitionsDefinition, self._partitions_def)
+        keys_per_dimension: Dict[str, Set] = defaultdict(set)
+        for partition_key in partition_keys:
+            for dimension, key in cast(MultiPartitionKey, partition_key).keys_by_dimension.items():
+                keys_per_dimension[dimension].add(key)
+
+        if set(keys_per_dimension.keys()) != set(
+            [dim.name for dim in partitions_def.partitions_defs]
+        ):
+            check.invariant(
+                "Provided partition keys have different dimensions than the partitions definition."
             )
+
+        for dim_def in cast(MultiPartitionsDefinition, partitions_def).partitions_defs:
+            if dim_def.name not in keys_per_dimension:
+                raise DagsterInvariantViolationError(
+                    f"Dimension {dim_def.name} not found in provided partition keys."
+                )
+
+            valid_keys = set(dim_def.partitions_def.get_partition_keys())
```

Comment:
This makes `with_partitions_key` a O(# partitions) an operation, which seems not ideal.